### PR TITLE
[feature] #2301: Add transaction's block hash when querying it

### LINKED ISSUE #2301: Add possibility to get transaction's block hash when querying it
## Motivation

In the Iroha Explorer, we fetch transactions with `FindAllTransactions` query, and we want to know to what block each of them belongs to.

## Possible solutions

- **(the most preferable)** In `FindAllTransactions` query return not `Vec<TransactionValue>` but some other struct like:

  ```rust
  struct TransactionExt {
    tx: TransactionValue,
    block_hash: Hash
  }
  ```

- Extend `Transaction` or `RejectedTransaction`  structures with `block_hash` field
- Extend general `TransactionValue` with `block_hash` field (i.e. make it struct instead of enum)

### PR DESCRIPTION
**Pull request by omkar-mohanty**
Signed-off-by: omkar-mohanty <franzohouser@gmail.com>

### Description of the Change
This PR introduces `TransactionQueryResult` struct. This struct holds the data of individual transaction as well as the hash of the block to which the transaction belongs to.
### Issue

Closes #2301 
In the Iroha Explorer, we fetch transactions with FindAllTransactions query, and we want to know to what block each of them belongs to.
### Benefits
We know which transaction belongs to which block.
### Possible Drawbacks
Additional struct to maintain  
### Alternate Designs *[optional]*
As proposed by @0x009922 
1. Extend `Transaction` or `RejectedTransaction` structures with `block_hash` field 
2. Extend general TransactionValue with block_hash field (i.e. make it struct instead of enum)
#### Possible Drawbacks as seen by  @[appetrosyan](https://github.com/appetrosyan)

> Extend `Transaction` or `RejectedTransaction` structures with `block_hash` field 

This could lead to an OOM attack

> Extend general TransactionValue with block_hash field (i.e. make it struct instead of enum) 

Problematic in terms of memory

### CODE REVIEW

File: data_model/src/transaction.rs [hunk 1]
```diff
@@ -439,6 +439,42 @@ impl PartialOrd for TransactionValue {
...
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.payload()
+            .creation_time
+            .cmp(&other.payload().creation_time)
+    }
+}
+
+impl PartialOrd for TransactionExt {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
```
Comment (appetrosyan) [inline data_model/src/transaction.rs:469]: Also `#[inline]`. This seems like a lot of syntax, but the actual assembly and MIR are quite small, so you win by inlining this comparison.

File: data_model/src/transaction.rs [hunk 2]
```diff
@@ -439,6 +439,43 @@ impl PartialOrd for TransactionValue {
...
+pub struct TransactionQueryResult {
+    /// Transaction
+    pub tx_value: TransactionValue,
+    /// Hash of the block to which `tx` belongs to
+    pub block_hash: Hash,
+}
+
+impl TransactionQueryResult {
+    #[inline]
+    /// Used to return payload of the transaction
```
Comment (outoftardis) [inline data_model/src/transaction.rs:453]:
```suggestion
    /// Return the payload of the transaction
```

File: data_model/src/transaction.rs [hunk 3]
```diff
@@ -439,6 +439,43 @@ impl PartialOrd for TransactionValue {
     }
 }
 
+/// `TransactionExt` is used in `FindAllTransactions` query
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub struct TransactionQueryResult {
+    /// Transaction
+    pub tx_value: TransactionValue,
+    /// Hash of the block to which `tx` belongs to
```
Comment (outoftardis) [inline data_model/src/transaction.rs:447]:
```suggestion
    /// The hash of the block to which `tx` belongs to
```

File: data_model/src/transaction.rs [hunk 4]
```diff
@@ -439,6 +439,43 @@ impl PartialOrd for TransactionValue {
...
+    }
+}
+
+impl PartialOrd for TransactionQueryResult {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(
+            self.payload()
+                .creation_time
+                .cmp(&other.payload().creation_time),
```
Comment (Arjentix) [inline data_model/src/transaction.rs:474]:
```suggestion
            self.cmp(other)
```

**Review by Arjentix — DISMISSED** [review]
Looks fine for me. Need to be rebased though

**Comment by Arjentix** [comment]
For the future contributions (I hope there will be some :)), please, leave a link to the issue in the `Issue` section of PR description. We usually write `Closes #2301`

**Review by appetrosyan — DISMISSED** [review]

**Comment by omkar-mohanty** [comment]
> Issue

Thanks for letting me know i will keep on contributing and next time I will write `closes` issue

File: core/src/smartcontracts/isi/asset.rs [hunk 5]
```diff
@@ -13,261 +16,317 @@ use super::prelude::*;
...
+
+                Ok(AssetEvent::Added(transfer.destination_id.clone()))
+            })?;
+            Ok(())
         }
     }
 
-    impl<W: WorldTrait> Execute<W> for RemoveKeyValue<Asset, Name> {
-        type Error = Error;
+    /// Trait for collecting fields associated with asset isi instructions:
```
Comment (outoftardis) [inline core/src/smartcontracts/isi/asset.rs:271]:
```suggestion
    /// Trait for collecting fields associated with Iroha Special Instructions for assets:
```

File: core/src/smartcontracts/isi/asset.rs [hunk 6]
```diff
@@ -268,7 +268,7 @@ pub mod isi {
         }
     }
 
-    /// Trait for collecting fields associated with asset isi instructions:
+    /// Trait for collecting fields associated with Iroha Special Insttructions for assets:
```
Comment (appetrosyan) [inline core/src/smartcontracts/isi/asset.rs:271]:
```suggestion
    /// Trait for collecting fields associated with Iroha Special Instructions for assets:
```

**Review by appetrosyan — APPROVED** [review]
Overall good.

**Review by appetrosyan — APPROVED** [review]

**Review by appetrosyan — DISMISSED** [review]

File: core/src/smartcontracts/isi/asset.rs (hunk 6 above)
Comment (s8sato) [reply core/src/smartcontracts/isi/asset.rs:271]: Could you address this @omkar-mohanty ?

**Review by s8sato — DISMISSED** [review]

**Comment by omkar-mohanty** [comment]
@s8sato I addressed the suggestions.

**Review by s8sato — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**